core/src/main/scala/caliban/validation/Validator.scala

```diff
@@ -545,61 +546,117 @@ object Validator {
+        case LIST | NON_NULL              => t.ofType.fold[Either[__Type, Unit]](Left(t))(isInputType)
+        case SCALAR | ENUM | INPUT_OBJECT => Right(())
+        case _                            => Left(t)
+      }
+    }
+
+    IO.whenCase(isInputType(`type`)) {
+      case Left(errorType) =>
         failValidation(
-          s"InputObject can't start with '__': ${field.name}",
-          """The input field must not have a name which begins with the
-characters {"__"} (two underscores)"""
+          s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an InputType",
+          """The input field must accept a type where IsInputType(type) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
         )
-      )
+    }
+  }
 
-    def onlyInputFieldType(field: __InputValue): IO[ValidationError, Unit] =
-      IO.whenCase(isInputType(field.`type`())) {
-        case Left(errorType) =>
-          failValidation(
-            s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an InputType",
-            """The input field must accept a type where IsInputType(inputFieldType) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
-          )
+  private def onlyOutputType(`type`: __Type): IO[ValidationError, Unit] = {
+    // https://spec.graphql.org/June2018/#IsOutputType()
+    def isOutputType(t: __Type): Either[__Type, Unit] = {
+      import __TypeKind._
+      t.kind match {
+        case LIST | NON_NULL                            => t.ofType.fold[Either[__Type, Unit]](Left(t))(isOutputType)
+        case SCALAR | OBJECT | INTERFACE | UNION | ENUM => Right(())
+        case _                                          => Left(t)
       }
+    }
 
-    t.inputFields match {
-      case None | Some(Nil) =>
+    IO.whenCase(isOutputType(`type`)) {
+      case Left(errorType) =>
         failValidation(
-          s"InputObject ${t.name.getOrElse("")} does not have fields",
-          "An Input Object type must define one or more input fields"
+          s"${errorType.name.getOrElse("")} is of kind ${errorType.kind}, must be an OutputType",
+          """The input field must accept a type where IsOutputType(type) returns true, https://spec.graphql.org/June2018/#IsInputType()"""
         )
-      case Some(fields) => validateFields(fields)
     }
   }
 
+  private def noDuplicateFieldName(fields: List[__Field]) =
+    noDuplicateName[__Field](fields, _.name)
+
+  private def noDuplicateInputValueName(inputValues: List[__InputValue]) =
+    noDuplicateName[__InputValue](inputValues, _.name)
+
+  private def noDuplicateName[T](listOfNamed: List[T], nameExtractor: T => String): IO[ValidationError, Unit] =
+    listOfNamed
+      .groupBy(nameExtractor(_))
+      .collectFirst { case (_, f :: _ :: _) => f }
+      .fold[IO[ValidationError, Unit]](IO.unit)(duplicate =>
+        failValidation(
+          s"InputObject has repeated fields: ${nameExtractor(duplicate)}",
```

Comment:
The message here seems okay for input objects but not for interfaces.

Reply:
added arguments `messageBuilder: T=> String` and `explanatoryText: String` for the caller to build the messages